### PR TITLE
feat: Allow notebook to mount feast-store config

### PR DESCRIPTION
**Pull request by harshad16**
feat: Allow notebook to mount feast-store config
Signed-off-by: Harshad Reddy Nalla <hnalla@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

## Description

Related-to: https://issues.redhat.com/browse/RHOAIENG-39802
This feature enables mounting of configmap created by feast store and via setting the info through label on notebook CR.
```
metadata:
  labels:
    opendatahub.io/feast-integration: true
```

## How Has This Been Tested?

```
make test
```
## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

## Summary by CodeRabbit

* **New Features**
  * Notebooks detect a Feast enablement label and mount a read-only Feast configuration from a ConfigMap into the matching notebook container; automatic creation/attachment of the config is handled.

* **Bug Fixes**
  * Injection failures are logged non-fatally and do not block admission; clearer handling when the Feast config or target container is missing; config is removed when the label is disabled.

* **Tests**
  * Added comprehensive unit and integration tests covering label handling, mount/unmount flows, multi-container cases, state transitions, and error scenarios.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by jstourac** [comment]
Thank you for this, Harshad. I haven't tried this yet on a cluster. I've put couple of comments to the existing code, otherwise no objections from me for now.

File: components/odh-notebook-controller/controllers/notebook_feast_config.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	feastConfigMapSuffix  = "-feast-config"
+	feastConfigVolumeName = "odh-feast-config"
+	feastConfigMountPath  = "/opt/app-root/config/feast"
```
Comment (andyatmiami) [inline components/odh-notebook-controller/controllers/notebook_feast_config.go]:
```suggestion
	feastConfigMountPath  = "/opt/app-root/src/config/feast"
```

we **DO** want to mount this feast config volume **within** the `/opt/app-root/src` volume.

File: components/odh-notebook-controller/controllers/notebook_feast_config.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+// NewFeastConfig creates a new Feast config.
+func NewFeastConfig(ctx context.Context, cli client.Client, notebook *nbv1.Notebook, log logr.Logger) error {
+
+	// validate feast config annotation is same as feast config map name
+	feastConfigMapName := notebook.Name + feastConfigMapSuffix
+	if notebook.Annotations[feastAnnotationKey] != feastConfigMapName {
+		return fmt.Errorf("feast config annotation is not in valid format, i.e <notebook-name>-feast-config")
+	}
```
Comment (andyatmiami) [inline components/odh-notebook-controller/controllers/notebook_feast_config.go]: This should be the `label` we are comparing against.

File: components/odh-notebook-controller/controllers/notebook_feast_config_test.go
Comment (andyatmiami) [inline components/odh-notebook-controller/controllers/notebook_feast_config_test.go]: Given latest updates to https://issues.redhat.com/browse/RHOAIENG-39802 - I think we want the tests to handle these 3 "happy path" scenarios (in addition to any other "edge cases")
- notebook **DOES NOT** have `opendatahub.io/feast-integration` label (no-op)
- notebook **DOES** have `opendatahub.io/feast-integration` label with value of `'false'` (no-op)
- notebook **DOES** have `opendatahub.io/feast-integration` label with value of `'true'` (new awesomeness)

File: components/odh-notebook-controller/controllers/notebook_feast_config_test.go [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+					},
+				},
+			}
+
+			err := NewFeastConfig(ctx, cli, notebook, ctrl.Log.WithName("test"))
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(ContainSubstring("feast config map is empty"))
+
+			// Cleanup
+			Expect(cli.Delete(ctx, configMap)).To(Succeed())
```
[the remark this answers is not in the document]
Comment (harshad16) [reply components/odh-notebook-controller/controllers/notebook_feast_config_test.go]: Adjusted with new changes, 
now included check of mount of configmap after integration test.
thanks for the suggestions

File: components/odh-notebook-controller/controllers/notebook_feast_config.go (hunk 1 above)
Comment (harshad16) [reply components/odh-notebook-controller/controllers/notebook_feast_config.go]: Acknowledge, adjusting to the code to use `/opt/app-roo/src/feast-config`
once i do the full testing, will adjust to `/opt/app-root/src/config/feast`

File: components/odh-notebook-controller/controllers/notebook_feast_config.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+	feastConfigMap := make(map[string]string)
+	// get the Feast config map
+	feastConfigMapObj := &corev1.ConfigMap{}
+	err := cli.Get(ctx, types.NamespacedName{Name: feastConfigMapName, Namespace: notebook.Namespace}, feastConfigMapObj)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			log.Error(err, "Feast config map not found", "ConfigMap", feastConfigMapName)
```
[the remark this answers is not in the document]
Comment (harshad16) [reply components/odh-notebook-controller/controllers/notebook_feast_config.go]: great point, i have removed all log.error, and moved them to only info level and switched to log without fail for this feature.

**Comment by harshad16** [comment]
Ready for Review

verified with the following changes:
- On Creation of notebook with label: `opendatahub.io/feast-integration: true`
Attaches configmap as volume
- On update of notebook with label: `opendatahub.io/feast-integration: false` and restart of notebook
unmount configmap as volume

[Screencast from 2025-12-10 22-49-17.webm](https://github.com/user-attachments/assets/3e26f991-4489-4ac6-8ad8-4897e14ee384)
[Screencast from 2025-12-10 18-14-35.webm](https://github.com/user-attachments/assets/bfaaaedf-740f-4e9e-91d4-be5696b6732a)

**Comment by andyatmiami** [comment]
/lgtm
/approve

nice work on this @harshad16 

folks can see a in-depth review of my test plan here (RH VPN required)
- https://gitlab.cee.redhat.com/-/snippets/10714

There are a few edge case I think we need to work with the feast team on correcting - but the behavior of the notebook-controller itself is exactly as I would expect.

One "expected" but maybe not obvious/intuitive behavior here on NBC:
- the `odh-notebook-controller` has logic that will **BLOCK** a change in the manifest from restarting the running notebook.  this means that if/as feast-related labels are added/removed/updated - the notebook must be explicitly restarted for the change to take effect.  as this is intended behavior that predates Feast integration - seems to me that this should be the outcome here.
